### ISSUE TITLE
IPython crashes when using the help operator during Variable Assignment

### ISSUE THREAD

**Issue by talkdirty** [ipython/ipython#10464]
I'm experiencing this (minor) issue when using IPython in this environment:

```js
{'commit_hash': 'e6e8653f5',
 'commit_source': 'repository',
 'default_encoding': 'UTF-8',
 'ipython_path': '/tmp/ipython/IPython',
 'ipython_version': '6.0.0.dev',
 'os_name': 'posix',
 'platform': 'Linux-4.10.1-1-ARCH-x86_64-with-arch',
 'sys_executable': '/tmp/ipython/vent/bin/python',
 'sys_platform': 'linux',
 'sys_version': '3.6.0 (default, Jan 16 2017, 12:12:55) \n[GCC 6.3.1 20170109]'}

```

I did not test different OS Environments, but the bug is also reproducible in the `ipython2` client.

# Steps to reproduce
IPython crashes when the `?` help operator is used during variable assignment.
```bash
$ ipython
Python 3.6.0 (default, Jan 16 2017, 12:12:55) 
Type "copyright", "credits" or "license" for more information.

IPython 5.1.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: test = open?

In [2]: test = open?                                                                                                                                                                            
Traceback (most recent call last):
  File "/usr/bin/ipython", line 5, in <module>
    start_ipython()
  File "/usr/lib/python3.6/site-packages/IPython/__init__.py", line 119, in start_ipython
    return launch_new_instance(argv=argv, **kwargs)
  File "/usr/lib/python3.6/site-packages/traitlets/config/application.py", line 658, in launch_instance
    app.start()
  File "/usr/lib/python3.6/site-packages/IPython/terminal/ipapp.py", line 348, in start
    self.shell.mainloop()
  File "/usr/lib/python3.6/site-packages/IPython/terminal/interactiveshell.py", line 440, in mainloop
    self.interact()
  File "/usr/lib/python3.6/site-packages/IPython/terminal/interactiveshell.py", line 423, in interact
    code = self.prompt_for_code()
  File "/usr/lib/python3.6/site-packages/IPython/terminal/interactiveshell.py", line 333, in prompt_for_code
    pre_run=self.pre_prompt, reset_current_buffer=True)
  File "/usr/lib/python3.6/site-packages/prompt_toolkit/interface.py", line 408, in run
    self._pre_run(pre_run)
  File "/usr/lib/python3.6/site-packages/prompt_toolkit/interface.py", line 383, in _pre_run
    pre_run()
  File "/usr/lib/python3.6/site-packages/IPython/terminal/interactiveshell.py", line 410, in pre_prompt
    self.pt_cli.application.buffer.text = cast_unicode_py2(self.rl_next_input)
  File "/usr/lib/python3.6/site-packages/prompt_toolkit/buffer.py", line 371, in text
    assert self.cursor_position <= len(value)
AssertionError

If you suspect this is an IPython bug, please report it at:
    https://github.com/ipython/ipython/issues
or send an email to the mailing list at ipython-dev@scipy.org

You can print a more detailed traceback right now with "%tb", or use "%debug"
to interactively debug it.

Extra-detailed tracebacks for bug-reporting purposes can be enabled via:
    %config Application.verbose_crash=True
```
Upon entering `test = open?` in the shell, the pager will open as usual, but once the pager is exited by hitting `q`, IPython will crash with the AssertionError above.
# Expected Behavior
- Preferred:
I am resumed to 
```python
In [1]: test = open?$
```
upon hitting `q` in the pager, where `$` denotes the position of the cursor.
- Alternative:
IPython displays an Error Message saying that the `?` operator should not be used during variable assignment.

**Comment by takluyver** [comment]
Your system info shows IPython 6.0 dev, but the info when you launch IPython shows `IPython 5.1.0`. This is an issue that was fixed for IPython 5.3.

**Comment by Carreau** [comment]
Please See also #10450 #10418 #10381 and #10348. 

Thanks.
